### PR TITLE
Resolves #2437 - Catch CSRF Failure when DEBUG=False

### LINKED ISSUE #2437: Catch CSRF Failure when DEBUG=False
### As ...

Patti - Platform Admin

### I want ...

the 403 CSRF failure to not render the default Django failure html when DEBUG is set to False

### So that ...

When errors occur in Nautobot a user remains in a view that has the same look and feel as nautobot and not a generic view without styling.

### I know this is done when...

- When form is submitted without CSRF token AND DEBUG is set to False a stylized view is returned with the error.
- When form is submitted and fails CSRF validation AND DEBUG is set to False a stylized view is returned with the error.
- When form is submitted without CSRF token AND DEBUG is set to True the default Django view is returned.
- When form is submitted and fails CSRF validation AND DEBUG is set to True the default Django view is returned.

### Optional - Feature groups this request pertains to.

- [ ] Automation
- [ ] Circuits
- [ ] DCIM
- [ ] IPAM
- [ ] Misc (including Data Sources)
- [ ] Organization
- [ ] Plugins (and other Extensibility)
- [ ] Security (Secrets, etc)
- [ ] Image Management
- [X] UI/UX
- [ ] Documentation
- [ ] Other (not directly a platform feature)

### Database Changes

None

### External Dependencies

None

### PR DESCRIPTION
**Pull request by whitej6**
# Closes: #2437 
# What's Changed

- Changed settings.py to use new view for CSRF failure
- Added template for failure
- Added view to render template when DEBUG=False OR return upstream Django view when DEBUG=True
image

# TODO

- [ ] Explanation of Change(s)
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by whitej6** [comment]
One question, would we want this to be high level that a failure occurred OR add the failure reason to the template?

**Comment by bryanculver** [comment]
Could we instead direct the user back one page in history instead of the homepage?

**Comment by whitej6** [comment]
This doesn't redirect to the homepage, it just returns a rendered error for the same URL which is what we do for catching 403/404/500 currently. 

In all likelihood this is a scenario that should not arise but was discovered when a tool for a customer first would make an API call but the URI variable was not instantiated so the POST went to `/` instead of `/api/dcim/foo/bar/`. The returned HTML had me confused on which item in the stack was the culprit since I had assumed CSRF 403 failures were caught just like typical 403. This was made more confusing due to a Django app that was installed which injected a lot of JS which is not part of Nautobot. The word Nautobot I don't _think_ was in the HTML response either since it was the generic 403 CSRF response.

Only time I have seen this in the wild has been when doing view development and creating custom HTML forms without adding the token to the form.

File: nautobot/core/views/__init__.py [hunk 1]
```diff
@@ -196,6 +197,20 @@ def server_error(request, template_name=ERROR_500_TEMPLATE_NAME):
...
+
+    If Nautobot is set to DEBUG the default view for csrf_failure.
+    `403_csrf_failure.html` template name is used over `403_csrf.html` to account for
+    additional context that is required to render the inherited templates.
+    """
+    if settings.DEBUG:
+        return _csrf_failure(request, reason=reason)
+    t = loader.get_template(template_name)
+    context = {"settings": settings}
+    return HttpResponseForbidden(t.render(context), content_type="text/html")
```
Comment (jathanism) [inline nautobot/core/views/__init__.py:211]: `csrf_failure` accepts a `template_name` argument:

```suggestion
    return _csrf_failure(request, reason, template_name)
```

File: nautobot/core/views/__init__.py [hunk 2]
```diff
@@ -4,13 +4,14 @@
 
 from django.conf import settings
 from django.contrib.auth.mixins import AccessMixin
-from django.http import HttpResponseServerError, JsonResponse
+from django.http import HttpResponseServerError, JsonResponse, HttpResponseForbidden
```
Comment (jathanism) [inline nautobot/core/views/__init__.py:7]:
```suggestion
from django.http import HttpResponseServerError, JsonResponse
```

File: nautobot/core/templates/403_csrf_failure.html [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{% extends '40x.html' %}
+
+{% block title %}Invalid Form Submission{% endblock %}
+
+{% block icon %}<i class="mdi mdi-alert"></i>{% endblock %}
+
+{% block message %}
+    CSRF failure has occurred.
```
Comment (jathanism) [inline nautobot/core/templates/403_csrf_failure.html:8]: Is there not a message that comes from the exception that is raised that is possibly more descriptive here?

Also could we not overload  the `more` context variable used in the [default template](https://github.com/django/django/blob/main/django/views/csrf.py)? vs. introducing another template to manage?

**Review by jathanism — CHANGES_REQUESTED** [review]

File: nautobot/core/views/__init__.py (hunk 1 above)
Comment (whitej6) [reply nautobot/core/views/__init__.py:211]: But does not have our context, which we need settings for the template to render. Initially I just added the template which should have worked if it had access to the context. The upstream view does not allow for outside context.

File: nautobot/core/views/__init__.py (hunk 2 above)
Comment (whitej6) [reply nautobot/core/views/__init__.py:7]: See above response.

File: nautobot/core/templates/403_csrf_failure.html (hunk 3 above)
Comment (whitej6) [reply nautobot/core/templates/403_csrf_failure.html:8]: See above responses and my first comment. 

Happy to add reason to the output.

Comment (jathanism) [reply nautobot/core/templates/403_csrf_failure.html:8]: I think that would be a nice middle ground. to say CSRF failed and at least attempt to explain why unless we provide a generic reason like "Refresh the page and try again." or something to that effect.

Comment (whitej6) [reply nautobot/core/templates/403_csrf_failure.html]: Added in reason and generic boilerplate from the 500.html template

File: changes/2437.added [hunk 4]
```diff
@@ -0,0 +1 @@
+Catch 403 CSRF errors and return stylized nautobot view.
```
Comment (glennmatthews) [inline changes/2437.added:1]:
```suggestion
Added Nautobot-themed error page for handling 403 CSRF errors.
```

**Review by glennmatthews — APPROVED** [review]

**Review by jathanism — APPROVED** [review]

**Pull request merged**